ocs_ci/ocs/resources/mcg_bucket.py

```diff
@@ -0,0 +1,78 @@
+import logging
+from abc import ABC, abstractmethod
+
+from ocs_ci.ocs import constants
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.utility import templating
+from ocs_ci.utility.utils import run_mcg_cmd
+from tests.helpers import create_unique_resource_name, create_resource
+
+logger = logging.getLogger(name=__file__)
+
+
+class MCGBucket(ABC):
+    """
+    A class to represent an MCG cloud
+    """
+    mcg, name = (None,) * 2
+
+    def __init__(self, mcg, name):
+        """
+        Constructor of an MCG bucket
+        """
+        self.mcg = mcg
+        self.name = name
+        logger.info(f"Creating bucket: {self.name}")
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if type(other) == str:
+            return self.name == other
+        elif type(other) == MCGBucket:
+            return self.name == other.name
+
+    def delete(self):
+        logger.info(f"Deleting bucket: {self.name}")
+        self.internal_delete()
+
+    @abstractmethod
+    def internal_delete(self):
+        pass
+
+
+class S3Bucket(MCGBucket):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mcg.s3_resource.create_bucket(Bucket=self.name)
+
+    def internal_delete(self):
+        self.mcg.s3_resource.Bucket(self.name).object_versions.delete()
+        self.mcg.s3_resource.Bucket(self.name).delete()
+
+    def list_all_buckets(self):
+        return [bucket.name for bucket in self.mcg.s3_resource.buckets.all()]
+
+
+class OCBucket(MCGBucket):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        obc_data = templating.load_yaml_to_dict(constants.MCG_OBC_YAML)
+        if self.name is None:
+            self.name = create_unique_resource_name('oc', 'obc')
+        obc_data['metadata']['name'] = self.name
+        obc_data['spec']['bucketName'] = self.name
+        create_resource(**obc_data)
```

Comment:
You can keep the resource here - self.data = create_resource()

Reply:
I considered this option and preferred to leave it out - considering that the CLI doesn't return an object of any kind, and that S3 returns a different kind of object, I worried it might cause more confusion than help.

Do you think it's preferable to add a `self.data` field and implement if for S3 and OC?

Reply:
I think that eventually, the desired state is to handle object exactly like file and block (from the testing perspective of course) so if we will follow the same guidelines for S3, MCG, object, etc. it will be much easier to onboard more ppl to the object storage world and help with the testing

Reply:
I agree that it'd be best to make the object storage world as accessible and straight-forward as possible - yet I worry that making `self.data` accessible will do the opposite.
If the bucket is an S3Bucket - `self.data` would contain an s3.Bucket object,
In case of a CLIBucket - `self.data` wouldn't contain anything (there is no inherent data rather than the name which has its own field)
And in case of an OCBucket, `self.data` would contain an OCP object.
Since the API is unified behind the scenes, and the user is never even exposed to the different kinds of buckets, I think that it'll be very confusing to hold different kinds of objects in `self.data` while on the surface the buckets all appear the same. (That is, for the user, all buckets are just 'MCGBuckets'. For us, they can be more specific like S3, OC or CLI, but that's strictly behind the scenes.)
In addition, I didn't run into any issues that required exposing the data, thus I didn't implement it yet.

Would be glad to have a deeper discussion regarding this